h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -17,14 +17,81 @@
    *
    * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained. Only Frame and Model keys are accepted.
    */
-  public static void retain(final Key[] retainedKeys){
-    final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
-    retainedSet.addAll(Arrays.asList(retainedKeys));
+  public static void retain(final Key[] retainedKeys) {
+    final Set<Key> retainedKeysSet = new HashSet<>(retainedKeys.length);
+    retainedKeysSet.addAll(Arrays.asList(retainedKeys));
     // Frames and models have multiple nested keys. Those must be extracted and kept from deletion as well.
-    extractNestedKeys(retainedSet);
+    extractNestedKeys(retainedKeysSet);
+    Set<Value> removedKeys = collectUniqueKeys();
+    removeValues(removedKeys, retainedKeysSet);
+  }
+
+  /**
+   * Collects a {@link Set} of keys available cluster-wide
+   *
+   * @return
+   */
+  private static final Set<Value> collectUniqueKeys() {
+    final Value[] collectedValues = new CollectValuesTask()
+            .doAllNodes()
+            ._collectedValues;
+    final Set<Value> clusterValues = new HashSet<>(collectedValues.length);
+
+    for (final Value value : collectedValues) {
+      clusterValues.add(value);
+    }
+
+    return clusterValues;
+  }
+
+  /**
+   * Removes values from DKV, except for keys meant to be retained.
+   *
+   * @param removedValues Values to be removed
+   * @param retainedKeys  A {@link Set} of keys to be retained
+   */
+  private static void removeValues(final Set<Value> removedValues, final Set<Key> retainedKeys) {
+
+    for (final Value value : removedValues) {
+      if (retainedKeys.contains(value._key)) {
+        continue;
+      }
+      final Futures removalFutures = new Futures();
```

Comment:
This all deletes the metadata (container objects like Frame / Model) sequentially, however the contained objects are still deleted in parallel.

Reply:
The containing objects should delete distinct objects - problem is parallel or even sequential invocation of deletion of a Model with default remove_impl - as this might cause nasty errors.